### DISCUSSION TITLE
REST API

### DISCUSSION THREAD

**Discussion by jindream6128** [24tngus/CS_STUDY#41] [Network]
# 6회차 REST API

REST Ful과 REST API

1. REST ful : `아키텍처 스타일 그 자체`
    - Representational State Transfer
    - 웹 아키텍처 스타일 중 하나 → 이는 웹 리소스(예: 문서, 이미지, 데이터)를 다루기 위한 일관된 방법을 제공하는 아키텍처 스타일
2. REST API : `아키텍처 스타일을 따르는 웹 서비스 인터페이스`
    - RESTful 아키텍처를 기반으로 구축된 웹 서비스의 인터페이스
    - CRUD작업을 수행할 때 HTTP 메서드 사용
    

## 1. REST API란?

- `한마디로 말했을때 url만 보고도 바로 어떤 데이터를 요청하는지 확인 하는 것`
- `REST` = Representational State Transfer **기계와 기계가 규격화된 방식으로 인터넷이나, 웹을 통해서 통신할수 있도록 돕는 통신 규칙 이며, 웹의 통신 규약인HTTP 를 사용한다.**
- `API` = Application Program Interface**컴퓨터의 기능을 실행시키는 방법**

## 2. REST 구성 요소

1. 자원(Resource) : URI
    - 모든 자원에는 고유한 ID가 존재하고 이 자원은 Server에 존재
    - HTTP 처럼 URI로 자원을 구분
2. 행위(Verb) : HTTP Method
    - HTTP 프로토콜의 Method 사용
3. 표현(Representation of Resource)
    - 클라이언트의 요청에 따른 서버의 응답 (JSON, XML, TEXT 등등)

### 3-1 Uniform - Interface (인터페이스의 일관성)

- URI로 지정한 Resource에 대한 조작을 통일되고 한정적인 인터페이스로 수행
    - url 자원식별 → 자원은 url로 식별
    - 표현을 통한 자원조작 → url과 GET, DELETE 등 HTTP 표준 메서드 등을 통해 자원을 조회, 삭제등과 같은 정보가 담겨 있어야 한다.
    - self - descriptive messages → HTTP Header에 타입을 명시하고 각 메시지 들은 MIME types에 맞춰 표현한다. (.json 반환 → application/json 명시)
    - HATEOAS 구조(Hypermedia as the Engine of Application State) → 하이퍼링크에 따라 다른 페이지를 보여줘야 하며 데이터마다 어떤 URL에서 원했는지 명시 ( href, link 등과같은 속성에 데이터 URL 명시)

### 3-2 Stateless (무상태성)

- HTTP 프로토콜은 무상태 프로토콜 이므로, REST API역시 무상태성을 가진다.

### 3-3 Cacheable (캐시 처리 가능)

- HTTP 프로토콜을 사용하므로, 웹에서 사용하는 기존의 기능 사용가능.

### 3-4 Client - Server 구조

- 클라이언트와 서버가 서로 독립적인 구조를 가져야 한다.

### 3-5 Layered System (계층화)

- Client는 REST API 서버만 호출하고, REST API 서버는 다중 계층으로 구

REST API의 URI 규칙에는 → 기본적으로 6가지가 존재한다.

1. 동작은 HTTP 메소드 만으로 구성
2. 확장자는 표시 해서는 안된다.
3. 동사가 아닌 명사로 표기
4. 계층 적인 내용을 담아야 한다 . /집/아파트/전세
5. 대문자가 아닌 소문자만 사용하고, 길 경우 - 사용 `( _ 사용 자제)`
6. HTTP 응답 상태 코드를 사용 

## `장점`

1. HTTP 프로토콜의 인프라를 그대로 사용하므로 , 별도의 인프라구축이 필요 없다.
2. HTTP를 사용하는 모든 플랫폼에서 사용 가능
3. 범용성이 높다.
4. 의도하는 바를 쉽게 나타낼 수 있다.
5. 서버와 클라이언트의 역할을 명확하게 분리

## `단점`

1. 표준이 존재하지 않아, 정의가 필요
2. HTTP 메소드 형태가 제한적 이다.

**Comment by 24tngus** [comment]
확인했습니다

**Comment by berryberries** [comment]
확인했습니다!
